File: ruby/.fleetControl/generate-schema.rb
#!/usr/bin/env ruby
# Fleet Control Config Schema Generator — Ruby Agent
#
# The Ruby newrelic.yml stores all optional config as commented-out dot-notation
# keys (e.g. "# application_logging.forwarding.enabled: true"). This script
# extracts those comment-style definitions and builds a nested JSON Schema.
#
# Exit codes: 0 = no changes, 1 = schema changed
# Dependencies: stdlib only (psych, json, net/http)

require 'json'
require 'net/http'
require 'psych'
require 'uri'

GITHUB_RAW_URL = 'https://raw.githubusercontent.com/newrelic/newrelic-ruby-agent/main/newrelic.yml'

SCRIPT_DIR  = File.dirname(File.expand_path(__FILE__))
SCHEMA_DIR  = File.join(SCRIPT_DIR, 'schemas')
K8S_SCHEMA  = File.join(SCHEMA_DIR, 'config-k8s.json')

# Enum overrides derived from comment text
ENUM_OVERRIDES = {
  'log_level'                              => %w[error warn info debug],
  'application_logging.forwarding.log_level' => %w[debug info warn error fatal unknown],
  'transaction_tracer.record_sql'          => %w[off raw obfuscated],
  'security.mode'                          => %w[IAST RASP],
}.freeze

EXCLUDE_PREFIXES = %w[
  audit_log
  instrumentation
  disable_
  heroku
  process_host
  marshaller
  preflight
  utilization
].freeze

# ---------------------------------------------------------------------------
# Fetch
# ---------------------------------------------------------------------------
def load_newrelic_yml
  local = ENV['NEWRELIC_YML']
  if local
    puts "Reading local: #{local}"
    return File.read(local)
  end
  puts "Fetching from GitHub: #{GITHUB_RAW_URL}"
  Net::HTTP.get(URI(GITHUB_RAW_URL))
end

# ---------------------------------------------------------------------------
# Type inference from a YAML value string
# ---------------------------------------------------------------------------
def infer_type(value_str)
  return ['boolean', false] if value_str == 'false'
  return ['boolean', true]  if value_str == 'true'
  return ['array',   []]    if value_str =~ /\A\[/
  return ['object',  {}]    if value_str =~ /\A\{/
  return ['integer', value_str.to_i] if value_str =~ /\A-?\d+\z/
  return ['number',  value_str.to_f] if value_str =~ /\A-?\d+\.\d+\z/
  ['string', value_str.empty? ? nil : value_str]
end

# ---------------------------------------------------------------------------
# Extract all config keys from the common: section.
# Ruby's YAML has most options as commented dot-notation lines:
#   # application_logging.forwarding.enabled: true
# Real uncommented keys (license_key, app_name, log_level) are also captured.
# ---------------------------------------------------------------------------
def extract_all_keys(raw_text)
  in_common = false
  results   = {}      # dot_path => { type, default, description }
  pending   = []      # comment lines before the current key

  raw_text.each_line do |line|
    stripped = line.strip

    if stripped.start_with?('common:')
      in_common = true
      next
    end
    break if in_common && stripped =~ /\A(development|test|staging|production):/

    next unless in_common

    # Real uncommented key: "  key_name: value"
    real = stripped.match(/\A([a-z_][a-z0-9_]*)\s*:\s*(.*)\z/)
    if real && !stripped.start_with?('#')
      key  = real[1]
      val  = real[2].gsub(/['"]?<%=.*?%>['"]?/, 'placeholder').strip
      type, default_val = infer_type(val)
      results[key] = { 'type' => type, 'default' => default_val,
                       'description' => pending.join(' ') }
      pending = []
      next
    end

    # Commented-out config key: "# dot.notation.key: value"
    commented = stripped.match(/\A#\s+([a-z_][a-z0-9_.]*)\s*:\s*(.*)\z/)
    if commented
      key_path = commented[1]
      val_str  = commented[2].strip

      # Skip keys that look like prose ("# See: docs.newrelic.com") or excluded prefixes
      unless key_path.include?(' ') ||
             EXCLUDE_PREFIXES.any? { |p| key_path.start_with?(p) }
        type, default_val = infer_type(val_str)
        results[key_path] ||= { 'type' => type, 'default' => default_val,
                                 'description' => pending.join(' ') }
      end
      pending = []
      next
    end

    # Accumulate description from comment lines
    if stripped.start_with?('#')
      text = stripped.sub(/\A#+\s*/, '').strip
      pending << text unless text.empty? ||
                             text =~ /\AFor example/i ||
                             text =~ /\A-\s/  # bullet list items
    else
      pending = []
    end
  end

  results
end

# ---------------------------------------------------------------------------
# Build nested JSON Schema properties from flat dot-notation map
# ---------------------------------------------------------------------------
def set_nested(node, parts, info)
  key = parts[0]
  if parts.length == 1
    node[key] = info
  else
    # If this slot is already a leaf, don't overwrite
    node[key] = {} unless node[key].is_a?(Hash)
    set_nested(node[key], parts[1..], info) unless node[key].key?('type') && node[key]['type'] != 'object'
  end
end

def tree_to_schema(tree)
  props = {}
  tree.each do |key, val|
    next unless val.is_a?(Hash)

    if val.key?('type') && val['type'] != 'object'
      # Leaf property
      dot_path = key  # approximate — good enough for enum lookup
      if ENUM_OVERRIDES.key?(dot_path)
        prop = { 'type' => 'string', 'enum' => ENUM_OVERRIDES[dot_path] }
        prop['default'] = val['default'] if ENUM_OVERRIDES[dot_path].include?(val['default'])
      else
        prop = { 'type' => val['type'] }
        prop['default'] = val['default'] unless val['default'].nil? || val['default'] == ''
        prop['items']              = { 'type' => 'string' } if val['type'] == 'array'
        prop['additionalProperties'] = true                 if val['type'] == 'object'
      end
      prop['description'] = val['description'] if val['description'] && !val['description'].empty?
      props[key] = prop
    else
      # Nested section: val is a Hash of sub-keys
      nested = tree_to_schema(val)
      next if nested.empty?
      props[key] = {
        'type'                 => 'object',
        'properties'           => nested,
        'additionalProperties' => true,
      }
    end
  end
  props
end

def build_properties(flat_map)
  tree = {}
  flat_map.each do |dot_path, info|
    parts = dot_path.split('.')
    set_nested(tree, parts, info)
  end
  tree_to_schema(tree)
end

# ---------------------------------------------------------------------------
# Schema generation
# ---------------------------------------------------------------------------
def generate_schema(raw_text)
  flat   = extract_all_keys(raw_text)
  props  = build_properties(flat)

  props['license_key'] = {
    'type'        => 'string',
    'description' => 'New Relic license key. Binds agent data to your account.',
    'minLength'   => 1,
  }

  {
    '$schema'              => 'https://json-schema.org/draft/2020-12/schema',
    'title'                => 'New Relic Ruby Agent Configuration',
    'description'          => 'Fleet Control configuration schema for the New Relic Ruby agent. Generated from newrelic.yml.',
    'type'                 => 'object',
    'properties'           => props,
    'required'             => %w[license_key app_name],
    'additionalProperties' => true,
  }
end

# ---------------------------------------------------------------------------
# Diff / write helpers
# ---------------------------------------------------------------------------
def diff_schemas(old_s, new_s, path = '')
  changes = []
  old_p   = (old_s || {})['properties'] || {}
  new_p   = (new_s || {})['properties'] || {}
  (old_p.keys | new_p.keys).each do |key|
    child = path.empty? ? key : "#{path}.#{key}"
    if    !old_p.key?(key)                                             then changes << "  + added:   #{child}"
    elsif !new_p.key?(key)                                             then changes << "  - removed: #{child}"
    elsif old_p[key]['type'] == 'object' && new_p[key]['type'] == 'object'
          changes.concat diff_schemas(old_p[key], new_p[key], child)
    elsif old_p[key] != new_p[key]                                     then changes << "  ~ changed: #{child}"
    end
  end
  changes.sort
end

def load_existing(path)
  return {} unless File.exist?(path)
  JSON.parse(File.read(path))
rescue JSON::ParserError
  {}
end

require 'fileutils'
def write_schema(schema, path)
  FileUtils.mkdir_p(File.dirname(path))
  File.write(path, JSON.pretty_generate(schema) + "\n")
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
raw      = load_newrelic_yml
schema   = generate_schema(raw)
old_k8s  = load_existing(K8S_SCHEMA)
changes  = diff_schemas(old_k8s, schema)

write_schema(schema, K8S_SCHEMA)
puts "Wrote: #{K8S_SCHEMA}"

if old_k8s.empty?
  puts "\nFirst run — schema created."
  exit 0
elsif changes.any?
  puts "\nSchema changes (#{changes.size}):"
  changes.each { |c| puts c }
  exit 1
else
  puts "\nNo schema changes."
  exit 0
end

File: nodejs/.fleetControl/generate-schema.js
#!/usr/bin/env node
/**
 * Fleet Control Config Schema Generator — Node.js Agent
 *
 * Fetches lib/config/default.js, stubs its internal require() dependencies,
 * calls definition() to get the full config structure, and generates
 * JSON Schema Draft 2020-12.
 *
 * Exit codes: 0 = no changes, 1 = schema changed
 * Dependencies: Node.js stdlib only
 */

'use strict';

const fs    = require('fs');
const https = require('https');
const path  = require('path');
const vm    = require('vm');

const GITHUB_RAW_URL =
  'https://raw.githubusercontent.com/newrelic/node-newrelic/main/lib/config/default.js';

const SCRIPT_DIR  = __dirname;
const SCHEMA_DIR  = path.join(SCRIPT_DIR, 'schemas');
const K8S_SCHEMA  = path.join(SCHEMA_DIR, 'config-k8s.json');

// Top-level keys to exclude (internal / runtime-only)
const EXCLUDE_KEYS = new Set([
  'newrelic_home', 'host', 'port', 'proxy_host', 'proxy_port', 'proxy_user',
  'proxy_pass', 'proxy_scheme', 'certificates', 'ignored_params',
  'ssl', 'apdex_t', 'capture_params', 'encoding_key', 'cross_application_tracer',
  'rum', 'browser_key', 'browser_monitoring_key',
  'debug', 'logger', 'config_path', 'config_file_path',
  'serverless_mode', 'feature_flag', 'utilization',
  'heroku', 'worker_threads', 'grpc',
]);

// Enum overrides
const ENUM_OVERRIDES = {
  'logging.level':                    ['fatal', 'error', 'warn', 'info', 'debug', 'trace'],
  'transaction_tracer.record_sql':    ['off', 'raw', 'obfuscated'],
  'transaction_tracer.obfuscated_sql_fields': null,  // skip
};

// ---------------------------------------------------------------------------
// Typed formatter stubs — we identify them by __schemaType later
// ---------------------------------------------------------------------------
function makeTypedStub(schemaType) {
  const fn = (v) => v;
  fn.__schemaType = schemaType;
  return fn;
}

const FORMATTER_STUBS = {
  array:      makeTypedStub('array'),
  int:        makeTypedStub('integer'),
  float:      makeTypedStub('number'),
  boolean:    makeTypedStub('boolean'),
  object:     makeTypedStub('object'),
  objectList: makeTypedStub('object'),
  allowList:  makeTypedStub('array'),
  regex:      makeTypedStub('string'),
};

// ---------------------------------------------------------------------------
// Fetch helpers
// ---------------------------------------------------------------------------
function fetchUrl(url) {
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        return fetchUrl(res.headers.location).then(resolve).catch(reject);
      }
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
      res.on('error', reject);
    }).on('error', reject);
  });
}

function loadDefaultJs() {
  const local = process.env.NEWRELIC_JS;
  if (local) {
    console.log(`Reading local: ${local}`);
    return Promise.resolve(fs.readFileSync(local, 'utf8'));
  }
  console.log(`Fetching from GitHub: ${GITHUB_RAW_URL}`);
  return fetchUrl(GITHUB_RAW_URL);
}

// ---------------------------------------------------------------------------
// Run default.js in a vm sandbox with stubbed requires
// ---------------------------------------------------------------------------
function evalDefaultJs(src) {
  const mod = { exports: {} };

  const sandbox = {
    module:     mod,
    exports:    mod.exports,
    __dirname:  '/stub',
    __filename: '/stub/default.js',
    console,
    process:    { env: {}, version: process.version, platform: process.platform, cwd: () => '/stub' },
    require(id) {
      if (id === './formatters')                   return FORMATTER_STUBS;
      if (id === './build-instrumentation-config') return [];
      if (id === './samplers')                     return { config: [] };
      if (id === 'path')                           return require('path');
      if (id === 'os')                             return require('os');
      // Any other internal require → empty object
      return {};
    },
  };
  sandbox.exports = sandbox.module.exports;

  try {
    new vm.Script(src, { filename: 'default.js' }).runInNewContext(sandbox);
  } catch (e) {
    console.warn(`Warning: vm eval error: ${e.message}`);
  }

  return sandbox.module.exports;
}

// ---------------------------------------------------------------------------
// Extract JSDoc descriptions: maps key name → description text
// ---------------------------------------------------------------------------
function extractJsDocDescriptions(src) {
  const descriptions = {};
  // Match /** ... */ blocks followed by a property definition
  const docBlockRe = /\/\*\*([\s\S]*?)\*\/\s*([a-zA-Z_$][a-zA-Z0-9_$]*)\s*:/g;
  let m;
  while ((m = docBlockRe.exec(src)) !== null) {
    const comment = m[1].replace(/\n\s*\*\s?/g, ' ').trim();
    const key     = m[2];
    // Strip @param/@returns tags
    const desc = comment.replace(/@\w+[^\n]*/g, '').trim().replace(/\s+/g, ' ');
    if (desc) descriptions[key] = desc;
  }
  return descriptions;
}

// ---------------------------------------------------------------------------
// Infer schema type from a default value + optional formatter reference
// ---------------------------------------------------------------------------
function inferType(value, formatter) {
  if (formatter && formatter.__schemaType) return formatter.__schemaType;
  if (typeof value === 'boolean')           return 'boolean';
  if (Number.isInteger(value))              return 'integer';
  if (typeof value === 'number')            return 'number';
  if (Array.isArray(value))                 return 'array';
  if (value !== null && typeof value === 'object') return 'object';
  return 'string';
}

// ---------------------------------------------------------------------------
// Recursively build JSON Schema properties from the definition() result
// ---------------------------------------------------------------------------
function buildProperties(node, descriptions, keyPath) {
  const properties = {};

  for (const [key, entry] of Object.entries(node)) {
    const fullPath = keyPath ? `${keyPath}.${key}` : key;

    if (EXCLUDE_KEYS.has(key))                   continue;
    if (ENUM_OVERRIDES[fullPath] === null)        continue;

    const isLeaf = entry !== null &&
                   typeof entry === 'object' &&
                   !Array.isArray(entry) &&
                   'default' in entry;

    if (isLeaf) {
      // Leaf config entry: { default, formatter?, env? }
      const { default: def, formatter, env } = entry;
      const jsonType = ENUM_OVERRIDES[fullPath]
        ? 'string'
        : inferType(def, typeof formatter === 'function' ? formatter : null);

      let prop;
      if (ENUM_OVERRIDES[fullPath]) {
        prop = { type: 'string', enum: ENUM_OVERRIDES[fullPath] };
        if (ENUM_OVERRIDES[fullPath].includes(def)) prop.default = def;
      } else {
        prop = { type: jsonType };
        if (def !== null && def !== undefined && jsonType !== 'object') {
          if (Array.isArray(def) && def.length === 0) {
            prop.items = { type: 'string' };
          } else if (!Array.isArray(def)) {
            prop.default = def;
          }
        }
        if (jsonType === 'array')  prop.items              = prop.items || { type: 'string' };
        if (jsonType === 'object') prop.additionalProperties = true;
      }
      if (env)                    prop['x-env-var']  = env;
      if (descriptions[key])      prop.description   = descriptions[key];
      properties[key] = prop;

    } else if (entry !== null && typeof entry === 'object' && !Array.isArray(entry)) {
      // Nested config section
      const nested = buildProperties(entry, descriptions, fullPath);
      if (Object.keys(nested).length > 0) {
        properties[key] = {
          type: 'object',
          properties: nested,
          additionalProperties: true,
        };
      }
    } else {
      // Raw primitive default (e.g. license_key: '')
      const jsonType = inferType(entry, null);
      const prop = { type: jsonType };
      if (entry !== null && entry !== '' && entry !== undefined) prop.default = entry;
      if (descriptions[key]) prop.description = descriptions[key];
      properties[key] = prop;
    }
  }

  return properties;
}

// ---------------------------------------------------------------------------
// Diff helpers
// ---------------------------------------------------------------------------
function diffSchemas(oldS, newS, prefix) {
  const changes = [];
  const oldP = (oldS && oldS.properties) || {};
  const newP = (newS && newS.properties) || {};
  for (const key of new Set([...Object.keys(oldP), ...Object.keys(newP)])) {
    const child = prefix ? `${prefix}.${key}` : key;
    if      (!(key in oldP))                                           changes.push(`  + added:   ${child}`);
    else if (!(key in newP))                                           changes.push(`  - removed: ${child}`);
    else if (oldP[key].type === 'object' && newP[key].type === 'object') changes.push(...diffSchemas(oldP[key], newP[key], child));
    else if (JSON.stringify(oldP[key]) !== JSON.stringify(newP[key])) changes.push(`  ~ changed: ${child}`);
  }
  return changes.sort();
}

function loadExisting(p) {
  try { return JSON.parse(fs.readFileSync(p, 'utf8')); } catch { return {}; }
}

function writeSchema(schema, p) {
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, JSON.stringify(schema, null, 2) + '\n');
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
loadDefaultJs().then((src) => {
  const exported     = evalDefaultJs(src);
  const descriptions = extractJsDocDescriptions(src);

  // node-newrelic exports defaultConfig.definition as a function
  const definitionFn = exported.definition;
  if (typeof definitionFn !== 'function') {
    console.error('ERROR: could not find definition() in default.js');
    process.exit(2);
  }

  const config     = definitionFn();
  const properties = buildProperties(config, descriptions, '');

  // Ensure license_key is properly typed (it's a raw '' in the source)
  properties['license_key'] = {
    type:           'string',
    description:    'New Relic license key. Binds agent data to your account.',
    minLength:      1,
    'x-env-var':    'NEW_RELIC_LICENSE_KEY',
  };

  const schema = {
    $schema:              'https://json-schema.org/draft/2020-12/schema',
    title:                'New Relic Node.js Agent Configuration',
    description:          'Fleet Control configuration schema for the New Relic Node.js agent. Generated from lib/config/default.js.',
    type:                 'object',
    properties,
    required:             ['license_key', 'app_name'],
    additionalProperties: true,
  };

  const oldK8s   = loadExisting(K8S_SCHEMA);
  const changes  = diffSchemas(oldK8s, schema, '');

  writeSchema(schema, K8S_SCHEMA);
  console.log(`Wrote: ${K8S_SCHEMA}`);

  if (!Object.keys(oldK8s).length) {
    console.log('\nFirst run — schema created.');
    process.exit(0);
  } else if (changes.length) {
    console.log(`\nSchema changes (${changes.length}):`);
    changes.forEach(c => console.log(c));
    process.exit(1);
  } else {
    console.log('\nNo schema changes.');
    process.exit(0);
  }
}).catch((err) => {
  console.error(`ERROR: ${err.message}`);
  process.exit(2);
});
